File: src/main/java/net/sacredlabyrinth/phaed/dynmap/simpleclans/DynmapSimpleClansListener.java
package net.sacredlabyrinth.phaed.dynmap.simpleclans;

import net.sacredlabyrinth.phaed.dynmap.simpleclans.entries.KillEntry;
import net.sacredlabyrinth.phaed.simpleclans.ClanPlayer;
import org.bukkit.entity.Arrow;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.EntityDeathEvent;
import org.bukkit.event.player.PlayerJoinEvent;

public class DynmapSimpleClansListener implements Listener {

    private final DynmapSimpleClans plugin;

    public DynmapSimpleClansListener(DynmapSimpleClans plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        plugin.getPlayerManager().addEntry(event.getPlayer());
    }

    @EventHandler(priority = EventPriority.LOW)
    public void onEntityDeath(EntityDeathEvent event) {
        if (event.getEntity() instanceof Player) {
            Player victim = (Player) event.getEntity();
            Player attacker = null;

            // find attacker

            EntityDamageEvent lastDamageCause = victim.getLastDamageCause();

            if (lastDamageCause instanceof EntityDamageByEntityEvent) {
                EntityDamageByEntityEvent entityEvent = (EntityDamageByEntityEvent) lastDamageCause;

                if (entityEvent.getDamager() instanceof Player) {
                    attacker = (Player) entityEvent.getDamager();
                } else if (entityEvent.getDamager() instanceof Arrow) {
                    Arrow arrow = (Arrow) entityEvent.getDamager();

                    if (arrow.getShooter() instanceof Player) {
                        attacker = (Player) arrow.getShooter();
                    }
                }
            }

            if (attacker != null) {
                ClanPlayer acp = plugin.getClanManager().getCreateClanPlayer(attacker.getUniqueId());
                ClanPlayer vcp = plugin.getClanManager().getCreateClanPlayer(victim.getUniqueId());

                DynmapSimpleClans.getInstance().getKills().addKillEntry(new KillEntry(vcp, acp, victim.getLocation()));
            }
        }
    }
}

File: src/main/kotlin/net/sacredlabyrinth/phaed/dynmap/simpleclans/DynmapSimpleClansListener.kt
package net.sacredlabyrinth.phaed.dynmap.simpleclans

import net.sacredlabyrinth.phaed.simpleclans.Clan
import net.sacredlabyrinth.phaed.simpleclans.Kill
import net.sacredlabyrinth.phaed.simpleclans.events.*
import org.bukkit.Bukkit.getScheduler
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority.MONITOR
import org.bukkit.event.Listener
import java.time.LocalDateTime

class DynmapSimpleClansListener(val plugin: DynmapSimpleClans) : Listener {

    @EventHandler(priority = MONITOR, ignoreCancelled = true)
    fun onSetHome(event: PlayerHomeSetEvent) = upsertMarkers(event.clan)

    @EventHandler(priority = MONITOR, ignoreCancelled = true)
    fun onHomeClear(event: PlayerHomeClearEvent) = deleteMarkers(event.clan.tag)

    @EventHandler(priority = MONITOR, ignoreCancelled = true)
    fun onDisband(event: DisbandClanEvent) = deleteMarkers(event.clan.tag)

    @EventHandler(priority = MONITOR, ignoreCancelled = true)
    fun onModtag(event: TagChangeEvent) = upsertMarkers(event.clan)

    @EventHandler(priority = MONITOR, ignoreCancelled = true)
    fun onKill(event: AddKillEvent) =
        plugin.killsLayer?.createMarker(Kill(event.attacker, event.victim, LocalDateTime.now()))

    private fun deleteMarkers(clanTag: String) {
        plugin.homeLayer?.markerSet?.findMarker(clanTag)?.deleteMarker()
        plugin.landsLayer?.markerSet?.findAreaMarker(clanTag)?.deleteMarker()
    }

    private fun upsertMarkers(clan: Clan) {
        // Running the runnable on next tick to get the actual clan data
        getScheduler().runTask(plugin, Runnable {
            plugin.homeLayer?.upsertMarker(clan)
            plugin.landsLayer?.upsertMarker(clan)
        })
    }
}

